### ISSUE TITLE
Add a way to hide the preview

### ISSUE THREAD

**Issue by flukeout** [mozilla/thimble.mozilla.org#1902]
Right now, if I want to focus on my code or markup, I make the preview window really small. To do that I have to drag the edge of the preview to the right side until it won't get any smaller.

This sucks because...

* Resizing in this way takes a long time
* The resizing handle is small and hard to grab
* It doesn't fully get rid of the preview

I think we should be able to toggle the preview window in the same way as the file tree.

**Comment by humphd** [comment]
Does this parallel the show/hide feature for the filetree, or the fullscreen preview that we do?  I agree that it would be nice to have the ability to only show the editor.

As an aside, we have code in Bramble to allow the preview window to be detached completely, and live in another window.  We thought we'd do that so you could have a full window with your code, and another window with your preview.

**Comment by flukeout** [comment]
More closely related to how we handle the file tree right now. We can expand into a 'full screen editor' as another phase.

**Comment by humphd** [comment]
OK, sounds good.  Can you do some kind of mock-up of what you want it to look like?  I'll get a student on it.

**Comment by flukeout** [comment]
Will do

**Comment by flukeout** [comment]
Here we go - this is what it will look like when the Preview is visible. There's an arrow icon on the right side of the preview bar. Clicking it hides the preview and expands the editor to the full width.

![image](https://cloud.githubusercontent.com/assets/25212/24119467/82e7cd90-0d6e-11e7-8c04-d6506413fb13.png)

When it's hidden, the button appears on the right side of the Editor bar. Clicking it restores the Preview.

![image](https://cloud.githubusercontent.com/assets/25212/24119596/0ca35cc0-0d6f-11e7-9e5e-3d80728c591c.png)

**Comment by humphd** [comment]
Fixing this will require a new API in Bramble.  We have code in https://github.com/mozilla/brackets/blob/master/src/extensions/default/bramble/lib/UI.js that deals with `#second-pane` and we make it resiable via the `Resizer` module.  I think you can call https://github.com/mozilla/brackets/blob/master/src/utils/Resizer.js#L93-L104 or `.hide()` on it, or we can just do it with css.

Whoever fixes this, you'll need to add code that's very similar to how we do the hide/show sidebar API:

* https://github.com/mozilla/brackets/blob/master/src/bramble/client/main.js#L855-L861 and see other places in that code that deal with `sidebar`
* https://github.com/mozilla/brackets/blob/master/src/extensions/default/bramble/lib/RemoteCommandHandler.js#L68-L161

@cgsingh, @peiying16, and some other people know how this code works, and might want to take this, or help someone with it.

**Comment by peiying16** [comment]
Do you mean the preview page doesn't restore?

**Comment by humphd** [comment]
@peiying16 he wants us to add a button in Thimble that works like the one for the filetree, and hides the live preview completely (so you can focus on just writing code).  I only mentioned you because the code to do this is very familiar to you by now, and I thought you might like this bug.  No pressure if you don't want to do it.

**Comment by peiying16** [comment]
I'm working with another issue now as you know. And I also have an idea to add this button, but I don't want to take many issues at some time. Other students may be interested in this issue. If anyone needs help, feel free to ask me. : )

**Comment by shubham0794x** [comment]
@humphd ,  I would like to contribute in solving this issue

**Comment by humphd** [comment]
Sure, @shubham0794x, go for it.  See notes above and ask questions as you have them.

**Comment by shubham0794x** [comment]
@humphd  Sir , we need to add the required feature in https://github.com/mozilla/brackets/blob/master/src/extensions/default/bramble/lib/UI.js .
And Sir , what do you mean by "we can just do it with css." ?

**Comment by humphd** [comment]
Yes, in `UI.js` I think, as discussed above in https://github.com/mozilla/thimble.mozilla.org/issues/1902#issuecomment-287893542.  "do it with css" would mean getting the `#second-pane` element and using a class that does `display:none`.  But that's what the `.hide()` stuff does, I think.

**Comment by Rajat-dhyani** [comment]
hey @shubham0794x  still working on this? Mind if i take over?

**Comment by GauravV-02** [comment]
Hi @flukeout , I am a student at seneca college enrolled in open source course, looking for my first bug. If this bug is not assigned to someone, can I work on it? Thanks

**Comment by flukeout** [comment]
Hey @GauravV-02 - hey, go ahead! Let me know if you get stuck or have questions. Looks like there have been a few attempts in other PRs to solve this, so please have a look at those for some context and additional information.

**Comment by GauravV-02** [comment]
Thanks, I will

**Comment by GauravV-02** [comment]
When I run, npm run env. I am getting errors. Here is the file with errors
[2017-09-28T19_59_28_290Z-debug.log](https://github.com/mozilla/thimble.mozilla.org/files/1342193/2017-09-28T19_59_28_290Z-debug.log)

**Comment by flukeout** [comment]
Paging @gideonthomas

**Comment by humphd** [comment]
@GauravV-02 for sure you ran `npm install` before that, yes?  It looks like it can't find the module binary it needs.

**Comment by GauravV-02** [comment]
Thank you it worked, but in the instruction this command was before npm install.

**Comment by humphd** [comment]
The docs might be wrong about the order.  Should you file another bug to update them?  If they are out of order, that's a bug we should fix.

**Comment by gideonthomas** [comment]
Nice catch @GauravV-02! Yes, the `npm install` command should be before the `npm run env` command. Feel free to file a bug and/or fix it as well.

**Comment by humphd** [comment]
OK, here's a two part plan for solving this.  The first plan is to make changes in Brackets to add an API for doing this; the second is to use that API and add UI to Thimble.

Let's start with the Brackets side.  We need to modify the code in https://github.com/mozilla/brackets/blob/master/src/extensions/default/bramble/lib/iframe-browser.js.  There is already some code to `detachPreview` and `attachPreview`, which we don't use anymore.  However, these contain the bulk of what we need.  Start by renaming `detachPreview()` to `hide()` and `attachPreview()` to `show()`.  There's already a method in there named `show()` which should be renamed to something like `setLayout()`, which is what it actually does.

Next, the `show()` and `hide()` functions need to get rewritten to remove some of the code.  They can be very simple (NOTE: I haven't tested the code below, but this is my guess as to what you'll need to do):

```js
function hide() {
            Resizer.hide("#second-pane");
            $("#first-pane").addClass("expandEditor");
}
...
function show() {
        Resizer.show("#second-pane");
        $("#first-pane").removeClass("expandEditor");
}
```

Now we need a way to call this code from Thimble.  To do that, we have to add two new API methods to the [Bramble IFrame API](https://github.com/mozilla/brackets#bramble-instance-methods).

A good way to approach this is to follow the same pattern as the `showSidebar()` and `hideSidebar()` methods, which are basically identical to what we need here:

* First we need new functions on the `bramble` object https://github.com/mozilla/brackets/blob/master/src/bramble/client/main.js#L1008-L1014
* Second, we need some code to listen for these methods being called inside Brackets, like this: https://github.com/mozilla/brackets/blob/master/src/extensions/default/bramble/lib/RemoteCommandHandler.js#L143-L148.  You'll change this code to call the `iframe-browser` module's `show()` and `hide()` methods that you wrote above.

@GauravV-02 see if this all makes sense.  I can answer questions you have.  Also, you should probably file a bug on your Vagrant issue with Thimble, since it's blocking your work in general.

**Comment by GauravV-02** [comment]
@humphd I have made all the changes mentioned above. Getting this on web console
"[Extension] failed to load /src/extensions/default/bramble - Error: Module name "iframe-browser" has not been loaded yet for context: _
http://requirejs.org/docs/errors.html#notloaded"

There is some problem here in the way i am declaring the variable or maybe the path
var PreviewView        = brackets.getModule("iframe-browser");

and I am using the variable here
case "BRAMBLE_SHOW_PREVIEW":
            PreviewView.show();
            break;
        case "BRAMBLE_HIDE_PREVIEW":
            PreviewView.hide();
            break;

**Comment by humphd** [comment]
@GauravV-02 can you open a PR and show me your full code?  It's hard to debug without the context.

**Comment by GauravV-02** [comment]
git push origin
remote: Permission to GauravV-02/brackets.git denied to gverma3.
fatal: unable to access 'https://github.com/GauravV-02/brackets.git/': The requested URL returned error: 403

Getting this error trying to fix it

**Comment by pradeepgangwar** [comment]
@humphd is this bug still open? If yes I would like to solve it. :)

**Comment by humphd** [comment]
This isn't a great first bug.  @pradeepgangwar I've commented in another bug you flagged, which is likely better.

For reference, @GauravV-02 has done the work on the Brackets side for this (the API has landed).  We just need the Thimble side now.
